File: opensrp-chw-referral/src/main/java/org/smartregister/chw/referral/util/ReferralUtil.java
package org.smartregister.chw.referral.util;

import org.joda.time.DateTime;
import org.smartregister.chw.referral.ReferralLibrary;
import org.smartregister.chw.referral.domain.ReferralTask;
import org.smartregister.domain.Task;
import org.smartregister.repository.AllSharedPreferences;
import org.smartregister.repository.BaseRepository;

import java.util.UUID;

public class ReferralUtil {

    public static void createReferralTask(ReferralTask referralTask, AllSharedPreferences allSharedPreferences) {
        Task task = new Task();
        task.setIdentifier(UUID.randomUUID().toString());
        //TODO Implement plans remove hard coded plan (in 2020 road-map)
      /*  Iterator<String> iterator = ChwApplication.getInstance().getPlanDefinitionRepository()
                .findAllPlanDefinitionIds().iterator();
        if (iterator.hasNext()) {
            task.setPlanIdentifier(iterator.next());
        } else {

            Timber.e("No plans exist in the server");
        }*/
        task.setPlanIdentifier(Constants.REFERRAl.PLAN_ID);
        task.setGroupIdentifier(referralTask.getGroupId());
        task.setStatus(Task.TaskStatus.READY);
        task.setBusinessStatus(Constants.BUSINESS_STATUS.REFERRED);
        task.setPriority(3);
        task.setCode(Constants.REFERRAl.CODE);
        task.setDescription(referralTask.getReferralDescription());
        task.setFocus(referralTask.getFocus());
        task.setForEntity(referralTask.getEvent().getBaseEntityId());
        DateTime now = new DateTime();
        task.setExecutionStartDate(now);
        task.setReasonReference(referralTask.getEvent().getEventId());
        task.setAuthoredOn(now);
        task.setLastModified(now);
        task.setOwner(allSharedPreferences.fetchRegisteredANM());
        task.setSyncStatus(BaseRepository.TYPE_Created);
        task.setRequester(allSharedPreferences.getANMPreferredName(allSharedPreferences.fetchRegisteredANM()));
        task.setLocation(allSharedPreferences.fetchUserLocalityId(allSharedPreferences.fetchRegisteredANM()));
        ReferralLibrary.getInstance().getTaskRepository().addOrUpdate(task);
    }
}

File: opensrp-chw-referral/src/main/java/org/smartregister/chw/referral/util/ReferralUtil.kt
package org.smartregister.chw.referral.util

import org.joda.time.DateTime
import org.smartregister.chw.referral.ReferralLibrary
import org.smartregister.chw.referral.domain.ReferralTask
import org.smartregister.chw.referral.util.Constants.BusinessStatus
import org.smartregister.chw.referral.util.Constants.Referral
import org.smartregister.domain.Task
import org.smartregister.repository.AllSharedPreferences
import org.smartregister.repository.BaseRepository
import timber.log.Timber
import java.util.*

object ReferralUtil {
    @JvmStatic
    fun createReferralTask(
        referralTask: ReferralTask, allSharedPreferences: AllSharedPreferences
    ) {
        val task = Task().apply {
            identifier = UUID.randomUUID().toString()
            /* //TODO Implement plans remove hard coded plan (in 2020 road-map)
             val iterator = ReferralLibrary.getInstance().getPlanDefinitionRepository()
                 .findAllPlanDefinitionIds().iterator()
             if (iterator.hasNext()) {
                 planIdentifier = iterator.next()
             } else {
                 Timber.e("No plans exist in the server")
             }*/
            planIdentifier = Referral.PLAN_ID
            groupIdentifier = referralTask.groupId
            status = Task.TaskStatus.READY
            businessStatus = BusinessStatus.REFERRED
            priority = 3
            code = Referral.CODE
            description = referralTask.referralDescription
            focus = referralTask.focus
            forEntity = referralTask.event.baseEntityId
            val now = DateTime()
            executionStartDate = now
            authoredOn = now
            lastModified = now
            owner = allSharedPreferences.fetchRegisteredANM()
            syncStatus = BaseRepository.TYPE_Created
            requester =
                allSharedPreferences.getANMPreferredName(allSharedPreferences.fetchRegisteredANM())
            location =
                allSharedPreferences.fetchUserLocalityId(allSharedPreferences.fetchRegisteredANM())
        }
        ReferralLibrary.getInstance().taskRepository.addOrUpdate(task)
    }
}